lvmd/embed.go

```diff
@@ -0,0 +1,128 @@
+package lvmd
+
+import (
+	"context"
+
+	"github.com/topolvm/topolvm/lvmd/proto"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/status"
+	gproto "google.golang.org/protobuf/proto"
+)
+
+// NewEmbeddedServiceClients creates clients locally calling instead of using gRPC.
+func NewEmbeddedServiceClients(ctx context.Context, dcmapper *DeviceClassManager, ocmapper *LvcreateOptionClassManager) (
+	proto.LVServiceClient,
+	proto.VGServiceClient,
+) {
+	vgServiceServerInstance, notifier := NewVGService(dcmapper)
+	lvServiceServerInstance := NewLVService(dcmapper, ocmapper, notifier)
+	caller := &localCaller{
+		lvServiceServer: lvServiceServerInstance,
+		vgServiceServer: vgServiceServerInstance.(*vgService),
+	}
+	caller.vgWatch = &localWatch{ctx: ctx, watch: make(chan any)}
+	go caller.vgServiceServer.Watch(&proto.Empty{}, caller.vgWatch)
+	return caller, caller
+}
+
+// localCaller is a struct holding indirections to the local lvmd server.
+// It implements both the LVServiceClient and VGServiceClient interfaces.
+// It also includes a watch for redirecting the vg watch to the local caller.
+type localCaller struct {
+	lvServiceServer proto.LVServiceServer
+	vgServiceServer proto.VGServiceServer
+	vgWatch         *localWatch
+}
+
+// localWatch is a local implementation of the WatchClient and WatchServer that is used by vgService.
+// it uses an unbound blocking channel to send and receive messages.
+type localWatch struct {
```

Comment:
I wonder if it is reasonable for `localWatch` to implement both `WatchClient`(= `proto.VGService_WatchClient`) and `WatchServer`(= `proto.VGService_WatchServer`).

`VGService_WatchClient` and `VGService_WatchServer` have common functions to be implemented (e.g. `Context`, `SendMsg`, `RecvMsg`). Though the current implementations of both interfaces happen to coincide, I believe they are semantically different.

How do you think about splitting`localWatch` into `embeddedWatchClient` and `embeddedWatchServer`?

Reply:
Usually I would agree, however this implementation is a pretty special case imo. Since the watchClient does nothing else then redirecting all calls into a channel that the watchServer has to react to, both share one unidirectional channel and have to implement Send and Receive methods. If I were to split the implementations, I would have to duplicate the code just to have different structs and share the channel pointer between both of them because they both need access to it. 

That would result in a lot more boilerplate code that doesn't really have any effect except for duplication since Send/Receive methods behave exactly the same.

I collected all of the methods I would have to duplicate below
```go
type SendReceiver interface {
	Send(m *proto.WatchResponse) error
	Recv() (*proto.WatchResponse, error)
	SendMsg(m any) error
	RecvMsg(m any) error
	Context() context.Context
}
```
if we split, then I would have to implement these twice or introduce a shared struct again that kind of defeats the purpose of what you are asking.

If you still think its better for style, I can adopt it, but I would like to keep it as is for simplicity

Reply:
I added some additional method comments to explain what is stubbed and why so that it is easier to understand the codebase. Also cleaned up the interfaces a bit. I hope that is acceptable for you. If not feel free to nit again and I will adjust.